File: second course/PDOO/civitasJava/src/juegoTexto/VistaTextual.java
package juegoTexto;

import civitas.CivitasJuego;
import civitas.Diario;
import civitas.OperacionesJuego;
import civitas.SalidasCarcel;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;
import GUI.Respuestas;

class VistaTextual {
  
  CivitasJuego juegoModel; 
  int iGestion=-1;
  int iPropiedad=-1;
  private static String separador = "=====================";
  
  private Scanner in;
  
  VistaTextual () {
    in = new Scanner (System.in);
  }
  
  void mostrarEstado(String estado) {
    System.out.println (estado);
  }
              
  void pausa() {
    System.out.print ("Pulsa una tecla");
    in.nextLine();
  }

  int leeEntero (int max, String msg1, String msg2) {
    Boolean ok;
    String cadena;
    int numero = -1;
    do {
      System.out.print (msg1);
      cadena = in.nextLine();
      try {  
        numero = Integer.parseInt(cadena);
        ok = true;
      } catch (NumberFormatException e) { // No se ha introducido un entero
        System.out.println (msg2);
        ok = false;  
      }
      if (ok && (numero < 0 || numero >= max)) {
        System.out.println (msg2);
        ok = false;
      }
    } while (!ok);

    return numero;
  }

  int menu (String titulo, ArrayList<String> lista) {
    String tab = "  ";
    int opcion;
    System.out.println (titulo);
    for (int i = 0; i < lista.size(); i++) {
      System.out.println (tab+i+"-"+lista.get(i));
    }

    opcion = leeEntero(lista.size(),
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo");
    return opcion;
  }

  SalidasCarcel salirCarcel() {
    int opcion = menu ("Elige la forma para intentar salir de la carcel",
      new ArrayList<> (Arrays.asList("Pagando","Tirando el dado")));
    return (SalidasCarcel.values()[opcion]);
  }

  Respuestas comprar(){
		String titulo = "Quiere comprar la calle?";
		ArrayList<String> lista = new ArrayList<>();
		lista.add("SI");
		lista.add("NO");
		
		if(menu(titulo,lista) == 0)
			return Respuestas.SI;
		else
			return Respuestas.NO;
	}

  void gestionar(){
		String titulo = "Gestion";
		ArrayList<String> listaGestiones = new ArrayList<>();
		ArrayList<String> listaProp = new ArrayList<>();
		
		listaGestiones.add("Vender");
    listaGestiones.add("Hipotecar");
    listaGestiones.add("Cancelar Hipoteca");      
    listaGestiones.add("Construir Casa");
    listaGestiones.add("Construir Hotel");
    listaGestiones.add("Terminar");
		
		iGestion = menu(titulo, listaGestiones);
		
		for(int i = 0; i < juegoModel.getJugadorActual().getPropiedades().size();i++)
			listaProp.add(juegoModel.getJugadorActual().getPropiedades().get(i).getNombre());
		
		if(!juegoModel.getJugadorActual().getPropiedades().isEmpty() && listaGestiones.get(iGestion) != "Terminar")
			iPropiedad = menu("Propiedad:", listaProp);
		else if(listaGestiones.get(iGestion) != "Terminar")
			System.out.println("\n No posee ninguna propiedad");
	}
  
  public int getGestion(){
		return iGestion;
	}
  
  public int getPropiedad(){
		return iPropiedad;
	}
    

  void mostrarSiguienteOperacion(OperacionesJuego operacion){
		System.out.println("Siguiente operacion a realizar: " + operacion);
	}


  void mostrarEventos(){
		while(Diario.getInstance().eventosPendientes())
			System.out.println(Diario.getInstance().leerEvento());
	}
  
  public void setCivitasJuego(CivitasJuego civitas){ 
        juegoModel=civitas;
        //this.actualizarVista();
	}
  
  void actualizarVista(){
		System.out.println("----ESTADO JUGADOR----");
		System.out.println(juegoModel.getJugadorActual().toString());
		System.out.println("----ESTADO CASILLA----");
		System.out.println(juegoModel.getCasillaActual().toString());
	} 
}

File: second course/PDOO/civitasRuby/pkg/civitasRuby-0.0.1/lib/vista_textual.rb
#encoding:utf-8
require_relative 'operaciones_juego'
require_relative 'respuestas'
require_relative 'salidas_carcel'
require_relative 'gestiones_inmobiliarias'
require_relative 'diario'
require_relative 'civitas_juego'
require 'io/console'

module Civitas

  class Vista_textual
		
		def initialize
			@iGestion = -1
			@iPropiedad = -1
		end

    def mostrar_estado(estado)
      puts estado
    end

    
    def pausa
      print "Pulsa una tecla"
      STDIN.getch
      print "\n"
    end

    def lee_entero(max,msg1,msg2)
      ok = false
      begin
        print msg1
        cadena = gets.chomp
        begin
          if (cadena =~ /\A\d+\Z/)
            numero = cadena.to_i
            ok = true
          else
            raise IOError
          end
        rescue IOError
          puts msg2
        end
        if (ok)
          if (numero >= max)
            ok = false
          end
        end
      end while (!ok)

      return numero
    end



    def menu(titulo,lista)
      tab = "  "
      puts titulo
      index = 0
      lista.each { |l|
        puts tab+index.to_s+"-"+l
        index += 1
      }

      opcion = lee_entero(lista.length,
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo")
      return opcion
    end
    
    def comprar
			lista_respuestas = [Respuestas::SI, Respuestas::NO]
      opcion = menu("Quiere comprar la calle?", ["SI", "NO"])
      return lista_respuestas[opcion]
    end

    def gestionar			
			titulo = "Gestion"

			lista_gestiones_inmobiliarias = [Gestiones_inmobiliarias::VENDER, 
																			Gestiones_inmobiliarias::HIPOTECAR, 
																			Gestiones_inmobiliarias::CANCELAR_HIPOTECA, 
																			Gestiones_inmobiliarias::CONSTRUIR_CASA, 
																			Gestiones_inmobiliarias::CONSTRUIR_HOTEL,
																			Gestiones_inmobiliarias::TERMINAR]
																	
			opcion = menu(titulo, ["Vender", "Hipotecar","Cancelar hipoteca",
														"Construir casa","Construir hotel", "Terminar"])
			
			@iGestion = opcion
			
			lista_prop = Array.new()
			
			for i in (0 ... @juego.jugador_actual.propiedades.size())
				lista_prop << @juego.jugador_actual.propiedades[i].nombre()	
			end
			
			if(!@juego.jugador_actual.propiedades.empty? && lista_gestiones_inmobiliarias[@iGestion] != Gestiones_inmobiliarias::TERMINAR)
				@iPropiedad = menu("Propiedades:" , lista_prop)
			elsif(lista_gestiones_inmobiliarias[@iGestion] != Gestiones_inmobiliarias::TERMINAR)	
				puts "\n No posee ninguna propiedad"
			end
			
    end

    def gestion()
			return @iGestion
    end

    def propiedad()
			return @iPropiedad
    end

    def mostrar_siguiente_operacion(operacion)
			puts "Siguiente operacion a realizar: #{operacion}"
    end

    def mostrar_eventos
			while(Diario.instance.eventos_pendientes)
        puts Diario.instance.leer_evento
      end
    end

    def set_civitas_juego(civitas)
         @juego=civitas
         #actualizar_vista()
    end

    def actualizar_vista()
			puts "----ESTADO JUGADOR----"
      puts @juego.jugador_actual.to_string()
      puts "----ESTADO CASILLA----"
      puts @juego.casilla_actual.to_string()
    end
		
		def salir_carcel
      lista_salidas_carcel = [Salidas_carcel::PAGANDO, Salidas_carcel::TIRANDO]
      opcion = menu("Elige la forma para intentar salir de la cárcel", ["Pagando", "Tirando el dado"])
      return lista_salidas_carcel[opcion]
    end
    
  end

end
